webzfs perms issue on linux, ARC bug on FreeBSD and basic replication error.

I feel kind of dumb raising a perm issue on Linux when nobody else has, but I don't I think I made any mistakes in installing/running webzfs. 

All test were carried out in virtual machines using both virt-manager and incus.   Access was remote as per docs with ssh port forward, or local and login to webzfs was via a standard non-root account.  Apologies for the "end user" style  report.

A. Linux (debian 13 )
===================================
Issue 1:  When running via the provided systemd until file, the "webzfs user" is denied permissions for zfs actions.

Webzfs is installed as per instructions and the "webzfs" user created as expected:

```
root@deb13vm:~# su - webzfs
webzfs@deb13vm:~$ id
uid=987(webzfs) gid=987(webzfs) groups=987(webzfs),42(shadow)
webzfs@deb13vm:~$ pwd
/opt/webzfs
webzfs@deb13vm:~$ getent group webzfs
webzfs:x:987:
webzfs@deb13vm:~$ getent passwd webzfs
webzfs:x:987:987:WebZFS Service User:/opt/webzfs:/bin/bash
webzfs@deb13vm:~$ getent shadow webzfs
webzfs:!:20465::::::
webzfs@deb13vm:~$ sudo ll
[sudo] password for webzfs:
sudo: a password is required
webzfs@deb13vm:~$ sudo -l
Matching Defaults entries for webzfs on deb13vm:
    env_reset, mail_badpass, secure_path=/usr/local/sbin\:/usr/local/bin\:/usr/sbin\:/usr/bin\:/sbin\:/bin, use_pty

User webzfs may run the following commands on deb13vm:
    (ALL) NOPASSWD: /usr/sbin/zpool, /usr/sbin/zfs, /usr/sbin/zdb -l *, /usr/bin/zpool, /usr/bin/zfs, /usr/bin/zdb -l *, /sbin/zpool, /sbin/zfs, /sbin/zdb -l *
    (ALL) NOPASSWD: /usr/sbin/smartctl, /usr/bin/smartctl
    (ALL) NOPASSWD: /usr/bin/lsblk, /usr/bin/blkid
    (ALL) NOPASSWD: /usr/sbin/sanoid, /usr/sbin/syncoid, /usr/bin/sanoid, /usr/bin/syncoid, /usr/local/sbin/sanoid, /usr/local/sbin/syncoid
    (ALL) NOPASSWD: /usr/bin/crontab
webzfs@deb13vm:~$
```
zfs commands are successful when executed by the "webzfs" user via an interactive login:

```
webzfs@deb13vm:~$ sudo zfs create -v tank/another
create tank/another
webzfs@deb13vm:~$ sudo zfs destroy tank/another
webzfs@deb13vm:~$
```

zfs commands fail (silently) when executed via webzfs webui, as logged in /opt/webzfs/.config/webzfs/logs, e.g:

```
webzfs@deb13vm:~/.config/webzfs/logs$ tail -n 2 zfs*
2026-01-19 09:17:09 [ERROR] user=chris operation=create_dataset status=FAILED dataset=tank/another error="Failed to create dataset: cannot create 'tank/another': permission denied
"
webzfs@deb13vm:~/.config/webzfs/logs$
```
Runing webzfs as root is the obvious workaround.

Issue 2:  I don't know if replication is meant to work at the moment, but a simple local non-incremental rep between datasets fails with the on screen message. e.g:

```
Source
    tank/test@2026.01.17-09.05.51
Target
    tank/another

Receive failed: cannot receive new filesystem stream: destination 'tank/another' exists must specify -F to overwrite it

```
After manually executing the zfs send/receive at the CLI and creating a further snap of tank/test, an incremental replication via the webzfs UI also fails with the same message:

Snaps on the pool:

```
root@deb13vm:~# zfs list -t snap | grep manual
tank/another@manual-2026-01-19_10-49                          0B      -   112M  -
tank/test@manual-2026-01-19_10-49                             0B      -   112M  -
tank/test@manual-2026-01-19_10-50                             0B      -   112M  -
root@deb13vm:~#
```

Result:

```
Source
    tank/test@manual-2026-01-19_10-50
Target
    tank/another
Duration
    0.04 seconds
Error
    Receive failed: cannot receive new filesystem stream: destination 'tank/another' exists must specify -F to overwrite it
```

Additionally, clicking on "view history" generates an exception in the app (first few lines shown):

```

127.0.0.1:37410 - "GET /zfs/replication/history HTTP/1.1" 500
[2026-01-19 10:59:05 +0000] [1758] [ERROR] Exception in ASGI application
Traceback (most recent call last):
  File "/opt/webzfs/views/zfs_replication.py", line 707, in replication_history
    return templates.TemplateResponse(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~^
        "zfs/replication/history.jinja",
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 ....

```

B. FreeBSD 15
===============

1. Issue accessing ARC info in the webzfs UI when running webzfs via the provided service script.

The package zfs-stats is installed:

```
root@freebsd15:~ # which zfs-stats
/usr/local/bin/zfs-stats
root@freebsd15:~ #
```

1.1 Under "Home > Observability > ARC Summary" the following on-screen error message appears:

```
Error: zfs-stats command not found. Install with: pkg install zfs-stats

Install zfs-stats with: pkg install zfs-stats

Add to sudoers: webzfs ALL=(ALL) NOPASSWD: /usr/local/bin/zfs-stats
```

But zfs-stats is installed and the webzfs install script does not create a "webzfs user" account on FreeBSD.

1.2 Under "Home > ZFS Performance> ARC Statistics" the following on-screen error message appears:

```
zfs-stats command not found. Install with: pkg install zfs-stats
Note: Install zfs-stats with: pkg install zfs-stats
```

If you run webzfs  using ".run_sh" from the CLI as root , then the  issue/errors in displaying ARC does not occur !

1.3 Under "Home > Replication > ZFS Send/Receive" a basic local non-incremental replication fails in the same ways as linux above.

--- a/services/zfs_observability.py
+++ b/services/zfs_observability.py
@@ -259,6 +259,13 @@ class ZFSObservabilityService:
         Returns:
             Dictionary with ARC stats
         """
+        if is_freebsd():
+            return self._get_arc_summary_freebsd()
+        else:
+            return self._get_arc_summary_linux()
+    
+    def _get_arc_summary_linux(self) -> Dict[str, Any]:
+        """Get ARC stats from Linux /proc filesystem"""
         try:
             arcstats_path = Path('/proc/spl/kstat/zfs/arcstats')
             
@@ -279,6 +286,53 @@ class ZFSObservabilityService:
                             stats[name] = int(value)
                         except ValueError:
                             stats[name] = value
+            
+            # Calculate derived stats
+            if 'hits' in stats and 'misses' in stats:
+                total = stats['hits'] + stats['misses']
+                if total > 0:
+                    stats['hit_rate'] = (stats['hits'] / total) * 100
+            
+            # Format sizes
+            if 'size' in stats:
+                stats['size_human'] = self._format_bytes(stats['size'])
+            if 'c_max' in stats:
+                stats['c_max_human'] = self._format_bytes(stats['c_max'])
+            
+            return stats
+            
+        except Exception as e:
+            return {'error': f'Failed to read ARC stats: {str(e)}'}
+    
+    def _get_arc_summary_freebsd(self) -> Dict[str, Any]:
+        """Get ARC stats from FreeBSD sysctl"""
+        try:
+            # Use sysctl to get ARC stats on FreeBSD
+            result = subprocess.run(
+                ['sysctl', 'kstat.zfs.misc.arcstats'],
+                capture_output=True,
+                text=True,
+                check=False
+            )
+            
+            if result.returncode != 0:
+                return {'error': f'Failed to get ARC stats: {result.stderr}'}
+            
+            stats = {}
+            for line in result.stdout.strip().split('\n'):
+                if not line or ':' not in line:
+                    continue
+                
+                # Format: kstat.zfs.misc.arcstats.hits: 12345678
+                key, value = line.split(':', 1)
+                # Extract just the stat name (last part after arcstats.)
+                stat_name = key.strip().split('.')[-1]
+                value = value.strip()
+                
+                try:
+                    stats[stat_name] = int(value)
+                except ValueError:
+                    stats[stat_name] = value
             
             # Calculate derived stats
             if 'hits' in stats and 'misses' in stats:

--- a/services/zfs_performance.py
+++ b/services/zfs_performance.py
@@ -636,6 +636,13 @@ class ZFSPerformanceService:
     
     def _read_arc_stats(self) -> Dict[str, Any]:
         """Read current ARC statistics"""
+        if self.system == 'FreeBSD':
+            return self._read_arc_stats_freebsd()
+        else:
+            return self._read_arc_stats_linux()
+    
+    def _read_arc_stats_linux(self) -> Dict[str, Any]:
+        """Read ARC stats from Linux /proc filesystem"""
         try:
             arcstats_path = Path('/proc/spl/kstat/zfs/arcstats')
             
@@ -656,6 +663,47 @@ class ZFSPerformanceService:
                             stats[name] = int(value)
                         except ValueError:
                             stats[name] = value
+            
+            # Calculate derived metrics
+            if 'hits' in stats and 'misses' in stats:
+                total = stats['hits'] + stats['misses']
+                if total > 0:
+                    stats['hit_rate'] = (stats['hits'] / total) * 100
+                    stats['miss_rate'] = (stats['misses'] / total) * 100
+            
+            return stats
+            
+        except Exception as e:
+            return {'error': f'Failed to read ARC stats: {str(e)}'}
+    
+    def _read_arc_stats_freebsd(self) -> Dict[str, Any]:
+        """Read ARC stats from FreeBSD sysctl"""
+        try:
+            result = subprocess.run(
+                ['sysctl', 'kstat.zfs.misc.arcstats'],
+                capture_output=True,
+                text=True,
+                check=False
+            )
+            
+            if result.returncode != 0:
+                return {'error': f'Failed to get ARC stats: {result.stderr}'}
+            
+            stats = {}
+            for line in result.stdout.strip().split('\n'):
+                if not line or ':' not in line:
+                    continue
+                
+                # Format: kstat.zfs.misc.arcstats.hits: 12345678
+                key, value = line.split(':', 1)
+                # Extract just the stat name (last part after arcstats.)
+                stat_name = key.strip().split('.')[-1]
+                value = value.strip()
+                
+                try:
+                    stats[stat_name] = int(value)
+                except ValueError:
+                    stats[stat_name] = value
             
             # Calculate derived metrics
             if 'hits' in stats and 'misses' in stats:
@@ -700,10 +748,11 @@ class ZFSPerformanceService:
                 }
                 
             elif self.system == 'FreeBSD':
-                # zfs-stats requires sudo on FreeBSD
+                # On FreeBSD, webzfs runs as root, so no sudo needed
+                # Use full path to ensure it's found regardless of PATH environment
                 try:
                     result = subprocess.run(
-                        ['sudo', 'zfs-stats', '-A'],
+                        ['/usr/local/bin/zfs-stats', '-A'],
                         capture_output=True,
                         text=True,
                         timeout=30
@@ -718,7 +767,7 @@ class ZFSPerformanceService:
                     return {
                         'output': result.stdout,
                         'system': self.system,
-                        'command': 'sudo zfs-stats -A',
+                        'command': 'zfs-stats -A',
                         'timestamp': datetime.now().isoformat()
                     }
                     

--- a/templates/zfs/observability/arc_summary.jinja
+++ b/templates/zfs/observability/arc_summary.jinja
@@ -108,7 +108,6 @@
                     <div>
                         <strong class="font-semibold">Error:</strong> {{ raw_arcstats.error }}
                         <p class="text-sm mt-2">Install zfs-stats with: <code>pkg install zfs-stats</code></p>
-                        <p class="text-sm">Add to sudoers: <code>webzfs ALL=(ALL) NOPASSWD: /usr/local/bin/zfs-stats</code></p>
                     </div>
                 </div>
             </div>

